File: src/example_client_sse.ts
/* eslint-disable no-console */
/**
 * Connect to the MCP server using SSE transport and call a tool.
 * You need provide a path to URL to the server
 */

import { Client } from '@modelcontextprotocol/sdk/client/index.js';
import { SSEClientTransport } from '@modelcontextprotocol/sdk/client/sse.js';
import { CallToolResultSchema } from '@modelcontextprotocol/sdk/types.js';
import { EventSource } from 'eventsource';

const SERVER_URL = 'http://localhost:3001/sse';

// @ts-expect-error: EventSource may not be defined on globalThis
if (typeof globalThis.EventSource === 'undefined') {
    // @ts-expect-error: Assigning a property that TypeScript doesn't currently recognize on globalThis
    globalThis.EventSource = EventSource;
}

async function main(): Promise<void> {
    const transport = new SSEClientTransport(new URL(SERVER_URL)
    );
    const client = new Client(
        { name: 'example-client', version: '1.0.0' },
        { capabilities: {} },
    );

    try {
        // Connect to the MCP server
        await client.connect(transport);

        // List available tools
        const tools = await client.listTools();
        console.log('Available tools:', tools);

        // Call a tool
        console.log('Calling "search" tool ...');
        const results = await client.callTool(
            { name: 'search', arguments: { query: 'web browser for Anthropic' } },
            CallToolResultSchema,
        );
        console.log('Tool result:', JSON.stringify(results, null, 2));

    } catch (error: unknown) {
        if (error instanceof Error) {
            console.error('Error:', error.message);
            console.error(error.stack);
        } else {
            console.error('An unknown error occurred:', error);
        }
    }
}

await main();

File: src/example_client_sse.py
import asyncio
import os

from dotenv import load_dotenv
from mcp.client.session import ClientSession
from mcp.client.sse import sse_client

load_dotenv(dotenv_path="../.env")

headers = {"Authorization": f"Bearer {os.getenv('APIFY_API_TOKEN')}"}

async def run():

    async with sse_client(url="https://jiri-spilka--rag-web-browser-task.apify.actor/sse", timeout=60, headers=headers) as (read, write):
        async with ClientSession(read, write) as session:
            await session.initialize()

            tools = await session.list_tools()
            print("Available Tools:", tools, end="\n\n")

            result = await session.call_tool("rag-web-browser", { "query": "example.com", "maxResults": 3 })
            print("Tools Call Result:")

            for content in result.content:
                print(content)

asyncio.run(run())
